File: src/cmd/ad/ad.c
#include "../../main/C.h"
#include "../../stacks/C.h"
#include "../../scan/C.h"
#include "../../misc/C.h"
#include "../../main/machinedep.h"

#define bcopy(s, d, n) memcpy(d, s, n)

/* Table of constant values */

static int c__2 = 2;

void ad_()
{
    /* Initialized data */

    static int cmdsav = -1;
    static short jbknd[3] = { 100,25,10 };


    /* System generated locals */
    int i__1;

#include "ad.dc.h"

    /* Local variables */
    int isav, init;
    int nscan;
    int itest, kount;
    extern int nxtscn_();

/*  ncare=0 means add will add all scans regardless of baseline */
/*      or clouds or pointing problems */
/*  ncare=1,2,3 means that add is increasingly severe about using */
/*      dubious scans */
/*  ncare=1 is default in cmi */

/* Set up the scan number range to the old values.  Parse can't do this */
    *sn1n = *sn2n = curscn_.num;
    bcopy(curscn_.fil, sn1f, 3);
    bcopy(curscn_.fil, sn2f, 3);

#include "ad.ec.h"

    i__1 = scan_.ndfn - 3;
    bcopy(sn1f, scan_.datfn + i__1, 3);
    subscanReq = *ssn;
/* set up limits, always declaring that limits were set */
    scnlmt_(sn1n, sn2n, sn1f, sn2f, (int)*istep, 0);
/* set up backend */
    if (*lbetyp > 0) {
	scan_.mbknd = jbknd[*lbetyp - 1];
    } else {
	scan_.mbknd = *lbknd;
    }
/* open data file and initialize a few things */
    opend_();
    bcopy(use_.muse, use_.nuse, sizeof(use_.nuse));
    if (cmnd_.cmdsig != cmdsav) {
	cmdsav = cmnd_.cmdsig;
	init = 1;
    } else {
	init = 0;
    }
    kount = 0;
    nscan = *sn1n;
L100:
    curscn_.num = nscan;
    if(read_(&curscn_.num)) {
	if(*sn1n == *sn2n)
	    error("Scan %d is empty", curscn_.num);
	++kount;
	goto L150;
    }
    if (nscan == *sn1n) {
	isav = scan_.iobstp;
    } else if (scan_.iobstp != isav) {
	++kount;
	goto L150;
    }
/* this is the place to check icare stuff */
    if (scan_.ndat[33] + *ncare > 2) {
	++kount;
	goto L150;
    }
    if(!get_())
	error("Bad backend");
/* do we need to check for matching header? */
    if (*mhflg != 0 && stk_[1].nlist != 0 && stk_[0].nlist != 0) {
	shift_(&c__2, &itest);
	if (itest != 0) {
	    ++kount;
	    goto L150;
	}
    }
    cmbn_(&init);
    init = 0;
L150:
    nscan = nxtscn_();
    if (nscan > 0) {
	goto L100;
    }
    if (kount != 0) {
	printf("%d scans not added\n", kount);
    }
    stk_[0] = stk_[1];
    main_.pltd = 0;
} /* ad_ */

File: src/cmd/ad/ad.f
	subroutine ad
c
c  ncare=0 means add will add all scans regardless of baseline
c      or clouds or pointing problems
c  ncare=1,2,3 means that add is increasingly severe about using
c      dubious scans
c  ncare=1 is default in cmi
	implicit undefined(a-z)
	include '../../stacks/F.h'
	include '../../main/F.h'
	include '../../scan/F.h'
	include 'ad.df.h'

	integer isav,init,kount,itest,nscan
	automatic isav,init,kount,itest,nscan
	integer cmdsav
	integer nxtscn
	integer*2 jbknd(3)
	include 'ad.if.h'
	data cmdsav/-1/
	data jbknd/100,25,10/

c set up 'old' values which parse doesn't easily deal with
	data snt/4/
c	sn1f= dfn1(1:3)
c	sn1n=nprssc
	sn2n=nprssc
	lbknd=mbknd
c call parse
	include 'ad.ef.h'
c Put current file name end in place
	datfn(ndfn-2:ndfn)=sn1f
c set up limits, always declaring that limits were set
	call scnlmt(sn1n,sn2n,sn1f,sn2f,istep,0)

c set up backend
	if(lbetyp.gt.0) then
		mbknd=jbknd(lbetyp)
	else
		mbknd=lbknd
	endif
c open data file and initialize a few things
	call opend
	call move(muse,nuse,maxchn/8)
	if(cmdsig.ne.cmdsav) then
		cmdsav=cmdsig
		init=1
	else
		init=0
	endif
      kount=0
	nscan = sn1n

100	nprssc=nscan
	call read(nprssc)
	if(nscan.eq.sn1n) then
		isav=ndat(13)
	else if(ndat(13).ne.isav) then
		kount=kount+1
		go to 150
	endif
c this is the place to check icare stuff
	if(ndat(34)+ncare.gt.2) then
		kount=kount+1
		go to 150
	endif
	call get
c do we need to check for matching header?
	if(mhflg.ne.0.and.nlist2.ne.0.and.nlist1.ne.0) then
		call shift(2,itest)
		if(itest.ne.0) then
			kount=kount+1
			go to 150
		endif
	endif
	call cmbn(init)
	init=0
150	nscan = nxtscn()
	if(nscan.gt.0) goto 100
	if(kount.ne.0) write(6,'(i5" Scans not added")')kount
	call move(numpt2,numpt1,stklen)
	pltd=0
	end
